### PR TITLE
NXP-29516/NXP-30044: async blob digest computation (10.10)

### PR DESCRIPTION
**Pull request by efge**

### CODE REVIEW

File: addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/test/java/org/nuxeo/ecm/blob/s3/TestS3BlobStoreTracing.java [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+
+    protected void checkTrace(String filename) throws IOException {
+        List<String> expectedTrace = LogTracingHelper.readTrace("traces/" + filename);
+        List<String> actualTrace = logCaptureResult.getCaughtEventMessages();
+        Map<String, String> context = new HashMap<>();
+        try {
+            LogTracingHelper.assertEqualsLists(expectedTrace, actualTrace, context);
+        } catch (AssertionError e) {
+            System.err.println(filename);
+            System.err.println(String.join("\n", actualTrace));
```
Comment (kevinleturc) [inline addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/test/java/org/nuxeo/ecm/blob/s3/TestS3BlobStoreTracing.java:174]: To replace by a logger?

File: nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/blob/AbstractBlobStore.java [hunk 2]
```diff
@@ -101,6 +115,76 @@ public String writeBlob(BlobContext blobContext) throws IOException {
...
+        }
+        return writeBlobGeneric(blobWriteContext);
+    }
+
+    /**
+     * Writes the blob without using any store-to-store optimization.
+     *
+     * @since 11.5
+     */
+    protected abstract String writeBlobGeneric(BlobWriteContext blobWriteContext) throws IOException;
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/blob/AbstractBlobStore.java:132]: Should we put a default implementation for the backport?

File: nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/blob/TestComputeDigestHelper.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        // make doc1 point to doc2's blob
+        Listener.properties = null;
+        String oldDigest = session.replaceBlobDigest(doc1.getRef(), FOO_MD5, BAR_MD5, "newdigest");
+        assertEquals(FOO_MD5, oldDigest);
+        // check event
+        assertNotNull(Listener.properties);
+        assertEquals(FOO_MD5, Listener.properties.get("key"));
+        assertEquals(FOO_MD5, Listener.properties.get("digest"));
+        assertEquals(BAR_MD5, Listener.properties.get("newKey"));
+        assertEquals("newdigest", Listener.properties.get("newDigest"));
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/blob/TestComputeDigestHelper.java:130]: Maybe using `CapturingEventListener` here?

**Review by kevinleturc — APPROVED** [review]
Great work 👍.

File: addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/test/java/org/nuxeo/ecm/blob/s3/TestS3BlobStoreTracing.java (hunk 1 above)
Comment (efge) [reply addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/test/java/org/nuxeo/ecm/blob/s3/TestS3BlobStoreTracing.java:174]: It's only for when the assertion fails, it's easier for me to have it on stderr.

File: nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/blob/AbstractBlobStore.java (hunk 2 above)
Comment (efge) [reply nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/blob/AbstractBlobStore.java:132]: The new default implem `AbstractBlobStore.writeBlob(BlobWriteContext)` is the only one that calls `writeBlobGeneric`. And an old customer `AbstractBlobStore` subclass had to have an existing `writeBlob(BlobWriteContext)` (because it's in the interface), so any existing customer implementation will continue to work and never call `writeBlobGeneric`. So I think we're ok.

Comment (efge) [reply nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/blob/AbstractBlobStore.java:132]: Hm OTOH it's `abstract` so recompilation would fail... Ok I'll add a default implementation throwing something.

File: nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/blob/TestComputeDigestHelper.java (hunk 3 above)
Comment (efge) [reply nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/blob/TestComputeDigestHelper.java:130]: Good idea

**Comment by efge** [comment]
Updated previous commits to answer comments.
Updated previous commit to add a `maxThreads` to the `ComputeDigestWork` queue config.
Added a new commit for NXP-30043 for safe concurrent behavior (using kv store).

**Review by kevinleturc — DISMISSED** [review]

**Comment by efge** [comment]
A few fixups in previous commits:
- log the new event `blobDigestUpdated` in the audit with the proper extended info (and CoreEventConstants for the keys).
- added `digestAlgorithm` with getter/setter in `AbstractBlob`, and updated `BlobStoreBlobProvider.fixupDigest` to set it correctly (this is tested in `TestS3BlobStoreTracing.testCopyAsyncDigest`).

Added a commit for NXP-30119 for the Drive Server changes.

**Comment by efge** [comment]
(last force push is just a 10.10 rebase)

**Review by kevinleturc — APPROVED** [review]

**Comment by efge** [comment]
T&P ok https://qa2.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-fguillaume-10.10/142/

**Comment by efge** [comment]
(rebased)

**Comment by efge** [comment]
(fixed one Javadoc issue + rebased)

File: addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/test/java/org/nuxeo/ecm/blob/s3/TestS3BlobStoreNoCaching.java [hunk 4]
```diff
@@ -37,4 +38,9 @@ public void testFlags() {
         assertTrue(bs.getKeyStrategy().useDeDuplication());
     }
 
+    @Override
+    public void testBlobGetFile() {
+        new AssumptionViolatedException("S3BlobStore with no caching has no File");
```
Comment (aimiebell) [inline addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/test/java/org/nuxeo/ecm/blob/s3/TestS3BlobStoreNoCaching.java:43]: I detect that this code is problematic. According to the [Correctness (CORRECTNESS)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#correctness-correctness), [RV: Exception created and dropped rather than thrown (RV_EXCEPTION_NOT_THROWN)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#rv-exception-created-and-dropped-rather-than-thrown-rv-exception-not-thrown).
This code creates an exception (or error) object, but doesn't do anything with it. For example, something like
It was probably the intent of the programmer to throw the created exception:

Comment (efge) [reply addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/test/java/org/nuxeo/ecm/blob/s3/TestS3BlobStoreNoCaching.java:43]: @aimiebell thanks. Please indicate in your message that this is a bot, it's bad form hide this info.

**Comment by efge** [comment]
Added two commits to make path strategies accept more keys, and improve CachingBlobStore to better use path strategies (don't depend on LRUFileCache); with added caching-related tests.

**Comment by efge** [comment]
T&P ok https://qa2.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-fguillaume-10.10/144/

**Comment by efge** [comment]
New commit to disallow async digest on VCS.

**Comment by efge** [comment]
Last force pushes were all about tweaking unit tests corner cases.

**Comment by efge** [comment]
T&P ok https://qa2.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-fguillaume-10.10/150/

**Review by kevinleturc — APPROVED** [review]

**Comment by efge** [comment]
T&P ok https://qa2.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-fguillaume-10.10/152/
